Text/PrettyPrint/Compact/Core.hs

```diff
@@ -1,83 +1,117 @@
-{-# LANGUAGE ScopedTypeVariables, TypeSynonymInstances, FlexibleContexts, FlexibleInstances, GeneralizedNewtypeDeriving, ViewPatterns #-}
-module Text.PrettyPrint.Compact.Core(Layout(..),Document(..),Doc) where
+{-# LANGUAGE ScopedTypeVariables, TypeSynonymInstances, FlexibleContexts, FlexibleInstances, GeneralizedNewtypeDeriving, ViewPatterns, DeriveFunctor, DeriveFoldable, DeriveTraversable #-}
+module Text.PrettyPrint.Compact.Core(Layout(..),Render(..),Document(..),Doc,annotate) where
 
-import Data.List (sort,groupBy,intercalate,minimumBy)
+import Data.List (sortOn,groupBy,minimumBy)
 import Data.Function (on)
 import Data.Semigroup
 import Data.Sequence (singleton, Seq, viewl, viewr, ViewL(..), ViewR(..), (|>))
 import Data.String
 import Data.Foldable (toList)
 
-newtype L = L (Seq String) -- non-empty sequence
-  deriving (Eq,Ord,Show)
+-- | Annotated string. We keep annotated segments in a separate list.
+--
+-- First argument is length.
+--
+data AS a = AS !Int [(Maybe a, String)]
```

Comment:
I'll document the invariants

we need a list (I think, have to write a test), because `Semigroup L` instance concatenates the `AS` (previously `String`), and we have to preserve the individual piece annotations.

I'm not sure whether `List` or `Seq` or something else is better from performance stand point, but that's easy to change.

I think that absence of annotation is quite common, also makes possible to avoid cluttering ~everything with `Monoid a` constraint.

If we agree on these points, I'll add them as notes to the source.